# Order reviews dataset 
This dataset includes data about the reviews made by the customers.

After a customer purchases the product from Olist Store a seller gets notified to fulfill that order. Once the customer receives the product, or the estimated delivery date is due, the customer gets a satisfaction survey by email where he can give a note for the purchase experience and write down some comments.

## Initial Column Description


|**Column Title**|**review_id-> str**|**order_id-> str** |**review_score-> int** |**review_comment_title -> str**| **review_comment_message -> str**|**review_creation_date -> timestamp**|**review_answer_timestamp -> timestamp**|
|--|--|--|--|--|--|--|--|
|Description |Primary key - unique review identifier |Primary key - unique order identifier |Note ranging from 1 to 5 given by the customer on a satisfaction survey. |Comment title from the review left by the customer, in Portuguese. |Comment message from the review left by the customer, in Portuguese.| Shows the date in which the satisfaction survey was sent to the customer.|Shows satisfaction survey answer timestamp. |
|Example |8670d52e15e00043ae7de4c01cc2fe06 |b9bf720beb4ab3728760088589c62129 |4 |recomendo	 |aparelho eficiente. no site a marca do aparelho esta impresso como 3desinfector e ao chegar esta com. |2018-05-22 00:00:00 |2018-05-23 16:45:47 |

## Required Libraries

In [1]:
import pandas as pd

## Data Preprocessing


In [2]:
review_csv = '../../data/pre_interim/order_reviews_dataset.csv'
review_dataset = pd.read_csv(review_csv)

In [3]:
review_dataset

review_id                          order_id  \
0      7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1      80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2      228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3      e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4      f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
...                                 ...                               ...   
99219  574ed12dd733e5fa530cfd4bbf39d7c9  2a8c23fee101d4d5662fa670396eb8da   
99220  f3897127253a9592a73be9bdfdf4ed7a  22ec9f0669f784db00fa86d035cf8602   
99221  b3de70c89b1510c4cd3d0649fd302472  55d4004744368f5571d1f590031933e4   
99222  1adeb9d84d72fe4e337617733eb85149  7725825d039fc1f0ceb7635e3f7d9206   
99223  efe49f1d6f951dd88b51e6ccd4cc548f  90531360ecb1eec2a1fbb265a0db0508   

       review_score review_comment_title  \
0                 4                  NaN   
1                 5                  NaN   
2                 5                  NaN   
3                 5                  NaN   
4                 5                  NaN   
...             ...                  ...   
99219             5                  NaN   
99220             5                  NaN   
99221             5                  NaN   
99222             4                  NaN   
99223             1                  NaN   

                                  review_comment_message review_creation_date  \
0                                                    NaN  2018-01-18 00:00:00   
1                                                    NaN  2018-03-10 00:00:00   
2                                                    NaN  2018-02-17 00:00:00   
3                   recebi bem antes do prazo estipulado  2017-04-21 00:00:00   
4      parabens lojas lannister adorei comprar pela i...  2018-03-01 00:00:00   
...                                                  ...                  ...   
99219                                                NaN  2018-07-07 00:00:00   
99220                                                NaN  2017-12-09 00:00:00   
99221  excelente mochila entrega super rapida super r...  2018-03-22 00:00:00   
99222                                                NaN  2018-07-01 00:00:00   
99223  meu produto chegou e ja tenho que devolver poi...  2017-07-03 00:00:00   

      review_answer_timestamp  
0         2018-01-18 21:46:59  
1         2018-03-11 03:05:13  
2         2018-02-18 14:36:24  
3         2017-04-21 22:02:06  
4         2018-03-02 10:26:53  
...                       ...  
99219     2018-07-14 17:18:30  
99220     2017-12-11 20:06:42  
99221     2018-03-23 09:10:43  
99222     2018-07-02 12:59:13  
99223     2017-07-03 21:01:49  

[99224 rows x 7 columns]

In [4]:
review_dataset.rename(columns={'order_id':'order'}, inplace=True)
review_dataset

review_id                             order  \
0      7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1      80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2      228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3      e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4      f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
...                                 ...                               ...   
99219  574ed12dd733e5fa530cfd4bbf39d7c9  2a8c23fee101d4d5662fa670396eb8da   
99220  f3897127253a9592a73be9bdfdf4ed7a  22ec9f0669f784db00fa86d035cf8602   
99221  b3de70c89b1510c4cd3d0649fd302472  55d4004744368f5571d1f590031933e4   
99222  1adeb9d84d72fe4e337617733eb85149  7725825d039fc1f0ceb7635e3f7d9206   
99223  efe49f1d6f951dd88b51e6ccd4cc548f  90531360ecb1eec2a1fbb265a0db0508   

       review_score review_comment_title  \
0                 4                  NaN   
1                 5                  NaN   
2                 5                  NaN   
3                 5                  NaN   
4                 5                  NaN   
...             ...                  ...   
99219             5                  NaN   
99220             5                  NaN   
99221             5                  NaN   
99222             4                  NaN   
99223             1                  NaN   

                                  review_comment_message review_creation_date  \
0                                                    NaN  2018-01-18 00:00:00   
1                                                    NaN  2018-03-10 00:00:00   
2                                                    NaN  2018-02-17 00:00:00   
3                   recebi bem antes do prazo estipulado  2017-04-21 00:00:00   
4      parabens lojas lannister adorei comprar pela i...  2018-03-01 00:00:00   
...                                                  ...                  ...   
99219                                                NaN  2018-07-07 00:00:00   
99220                                                NaN  2017-12-09 00:00:00   
99221  excelente mochila entrega super rapida super r...  2018-03-22 00:00:00   
99222                                                NaN  2018-07-01 00:00:00   
99223  meu produto chegou e ja tenho que devolver poi...  2017-07-03 00:00:00   

      review_answer_timestamp  
0         2018-01-18 21:46:59  
1         2018-03-11 03:05:13  
2         2018-02-18 14:36:24  
3         2017-04-21 22:02:06  
4         2018-03-02 10:26:53  
...                       ...  
99219     2018-07-14 17:18:30  
99220     2017-12-11 20:06:42  
99221     2018-03-23 09:10:43  
99222     2018-07-02 12:59:13  
99223     2017-07-03 21:01:49  

[99224 rows x 7 columns]

In [5]:
order_id_csv = '../../data/interim/unique_order_id.csv'
order_id = pd.read_csv(order_id_csv, dtype={'order_id':'Int64'})
order_id

order_id                             order
0             1  b81ef226f3fe1789b1e8b2acac839d17
1             2  a9810da82917af2d9aefd1278f1dcfa0
2             3  25e8ea4e93396b6fa0d3dd708e76c1bd
3             4  ba78997921bbcdc1373bb41e913ab953
4             5  42fdf880ba16b47b59251dd489d4441a
...         ...                               ...
99435     99436  0406037ad97740d563a178ecc7a2075c
99436     99437  7b905861d7c825891d6347454ea7863f
99437     99438  32609bbb3dd69b3c066a6860554a77bf
99438     99439  b8b61059626efa996a60be9bb9320e10
99439     99440  28bbae6599b09d39ca406b747b6632b1

[99440 rows x 2 columns]

#### Making merge between `review_dataset`.csv and `order_id`.csv

In [6]:
review_database = pd.merge(review_dataset,order_id, on='order', how='left')
review_database.drop(columns=['order'], inplace=True)
review_database = review_database.reindex(columns=['review_id','review_score','review_comment_title','review_comment_message','review_creation_date','review_answer_timestamp','order_id'])
review_database

review_id  review_score review_comment_title  \
0      7bc2406110b926393aa56f80a40eba40             4                  NaN   
1      80e641a11e56f04c1ad469d5645fdfde             5                  NaN   
2      228ce5500dc1d8e020d8d1322874b6f0             5                  NaN   
3      e64fb393e7b32834bb789ff8bb30750e             5                  NaN   
4      f7c4243c7fe1938f181bec41a392bdeb             5                  NaN   
...                                 ...           ...                  ...   
99219  574ed12dd733e5fa530cfd4bbf39d7c9             5                  NaN   
99220  f3897127253a9592a73be9bdfdf4ed7a             5                  NaN   
99221  b3de70c89b1510c4cd3d0649fd302472             5                  NaN   
99222  1adeb9d84d72fe4e337617733eb85149             4                  NaN   
99223  efe49f1d6f951dd88b51e6ccd4cc548f             1                  NaN   

                                  review_comment_message review_creation_date  \
0                                                    NaN  2018-01-18 00:00:00   
1                                                    NaN  2018-03-10 00:00:00   
2                                                    NaN  2018-02-17 00:00:00   
3                   recebi bem antes do prazo estipulado  2017-04-21 00:00:00   
4      parabens lojas lannister adorei comprar pela i...  2018-03-01 00:00:00   
...                                                  ...                  ...   
99219                                                NaN  2018-07-07 00:00:00   
99220                                                NaN  2017-12-09 00:00:00   
99221  excelente mochila entrega super rapida super r...  2018-03-22 00:00:00   
99222                                                NaN  2018-07-01 00:00:00   
99223  meu produto chegou e ja tenho que devolver poi...  2017-07-03 00:00:00   

      review_answer_timestamp  order_id  
0         2018-01-18 21:46:59     34776  
1         2018-03-11 03:05:13     34678  
2         2018-02-18 14:36:24     94785  
3         2017-04-21 22:02:06      7562  
4         2018-03-02 10:26:53     83967  
...                       ...       ...  
99219     2018-07-14 17:18:30     45544  
99220     2017-12-11 20:06:42     79038  
99221     2018-03-23 09:10:43     68567  
99222     2018-07-02 12:59:13      2870  
99223     2017-07-03 21:01:49     25360  

[99224 rows x 7 columns]

In [7]:
review_database.describe(include='all')

review_id  review_score review_comment_title  \
count                              99224  99224.000000                11476   
unique                             98410           NaN                 3416   
top     7b606b0d57b078384f0b58eac1d41d78           NaN            recomendo   
freq                                   3           NaN                  911   
mean                                 NaN      4.086421                  NaN   
std                                  NaN      1.347579                  NaN   
min                                  NaN      1.000000                  NaN   
25%                                  NaN      4.000000                  NaN   
50%                                  NaN      5.000000                  NaN   
75%                                  NaN      5.000000                  NaN   
max                                  NaN      5.000000                  NaN   

       review_comment_message review_creation_date review_answer_timestamp  \
count                   40842                99224                   99224   
unique                  34127                  636                   98249   
top                 muito bom  2017-12-19 00:00:00     2017-06-15 23:21:05   
freq                      492                  463                       4   
mean                      NaN                  NaN                     NaN   
std                       NaN                  NaN                     NaN   
min                       NaN                  NaN                     NaN   
25%                       NaN                  NaN                     NaN   
50%                       NaN                  NaN                     NaN   
75%                       NaN                  NaN                     NaN   
max                       NaN                  NaN                     NaN   

            order_id  
count   99223.000000  
unique           NaN  
top              NaN  
freq             NaN  
mean    49702.583302  
std     28701.072871  
min         1.000000  
25%     24860.500000  
50%     49690.000000  
75%     74550.500000  
max     99440.000000

In [8]:
items = len(review_database)
review_id = range(1,items+1)
review_database.insert(0,'id',review_id)
review_database

id                         review_id  review_score  \
0          1  7bc2406110b926393aa56f80a40eba40             4   
1          2  80e641a11e56f04c1ad469d5645fdfde             5   
2          3  228ce5500dc1d8e020d8d1322874b6f0             5   
3          4  e64fb393e7b32834bb789ff8bb30750e             5   
4          5  f7c4243c7fe1938f181bec41a392bdeb             5   
...      ...                               ...           ...   
99219  99220  574ed12dd733e5fa530cfd4bbf39d7c9             5   
99220  99221  f3897127253a9592a73be9bdfdf4ed7a             5   
99221  99222  b3de70c89b1510c4cd3d0649fd302472             5   
99222  99223  1adeb9d84d72fe4e337617733eb85149             4   
99223  99224  efe49f1d6f951dd88b51e6ccd4cc548f             1   

      review_comment_title                             review_comment_message  \
0                      NaN                                                NaN   
1                      NaN                                                NaN   
2                      NaN                                                NaN   
3                      NaN               recebi bem antes do prazo estipulado   
4                      NaN  parabens lojas lannister adorei comprar pela i...   
...                    ...                                                ...   
99219                  NaN                                                NaN   
99220                  NaN                                                NaN   
99221                  NaN  excelente mochila entrega super rapida super r...   
99222                  NaN                                                NaN   
99223                  NaN  meu produto chegou e ja tenho que devolver poi...   

      review_creation_date review_answer_timestamp  order_id  
0      2018-01-18 00:00:00     2018-01-18 21:46:59     34776  
1      2018-03-10 00:00:00     2018-03-11 03:05:13     34678  
2      2018-02-17 00:00:00     2018-02-18 14:36:24     94785  
3      2017-04-21 00:00:00     2017-04-21 22:02:06      7562  
4      2018-03-01 00:00:00     2018-03-02 10:26:53     83967  
...                    ...                     ...       ...  
99219  2018-07-07 00:00:00     2018-07-14 17:18:30     45544  
99220  2017-12-09 00:00:00     2017-12-11 20:06:42     79038  
99221  2018-03-22 00:00:00     2018-03-23 09:10:43     68567  
99222  2018-07-01 00:00:00     2018-07-02 12:59:13      2870  
99223  2017-07-03 00:00:00     2017-07-03 21:01:49     25360  

[99224 rows x 8 columns]

#### Creatign the dataset `review_database.csv`

When you saved the dataset always mark **"index = False"**. Or pandas will add a new column with a consequtive number. This small script is to remove this useless column.

In [9]:
review_database.to_csv('../../data/interim/review_database.csv', index=False)

## Final Column Description

|**Column Title**|**id ->int**|**review_id-> str**|**order_id-> str** |**review_score-> int** |**review_comment_title -> str**| **review_comment_message -> str**|**review_creation_date -> timestamp**|**review_answer_timestamp -> timestamp**|**order_id -> int**|
|--|--|--|--|--|--|--|--|--|--|
|Before processing |1 |8670d52e15e00043ae7de4c01cc2fe06 |b9bf720beb4ab3728760088589c62129 |4 |recomendo	 |aparelho eficiente. no site a marca do aparelho esta impresso como 3desinfector e ao chegar esta com. |2018-05-22 00:00:00 |2018-05-23 16:45:47 |2870|